### PR TITLE
Feature/pipeline multi instance

### PR DESCRIPTION
**Pull request by chengjoey**
#### What type of this PR

Add one of the following kinds:
/kind feature

#### What this PR does / why we need it:
add leader election pkg
pipeline reconcile make leader election
scheduler manager listen etcd for cluster info event 

#### Which issue(s) this PR fixes:

- Fixes #your-issue_number
- [Erda Cloud Issue Link](https://erda-org.erda.cloud/erda/dop/projects/387/issues/all?id=207973&issueFilter__urlQuery=eyJpdGVyYXRpb25JRHMiOls1MDZdLCJzdGF0ZUJlbG9uZ3MiOlsiT1BFTiIsIldPUktJTkciLCJXT05URklYIiwiUkVPUEVOIl0sImFzc2lnbmVlSURzIjpbIjEwMDEyMDUiXX0%3D&issueTable__urlQuery=eyJwYWdlTm8iOjF9&issueViewGroup__urlQuery=eyJ2YWx1ZSI6ImthbmJhbiIsImNoaWxkcmVuVmFsdWUiOnsia2FuYmFuIjoiZGVhZGxpbmUifX0%3D&iterationID=506&type=TASK)

#### Specified Reviewers:

/assign @your-reviewer

#### ChangeLog

| Language | Changelog |
| --------- | ------------ |
| 🇺🇸 English |   pipeline multi instance           |
| 🇨🇳 中文    | pipeline 单点修复             |

### CODE REVIEW

**Comment by sfwn** [comment]
Polish your commits, there are two commits with same msg.

File: modules/pipeline/pkg/leaderelection/leaderlection.go [hunk 1]
```diff
@@ -0,0 +1,60 @@
+// Copyright (c) 2021 Terminus, Inc.
```
Comment (sfwn) [inline modules/pipeline/pkg/leaderelection/leaderlection.go:1]: use https://github.com/erda-project/erda-infra/blob/master/providers/etcd-election/election.go.

**Review by sfwn — CHANGES_REQUESTED** [review]

File: conf/pipeline/pipeline.yaml [hunk 2]
```diff
@@ -14,6 +14,19 @@ mysql-xorm:
...
+  tls:
+    cert_file: "${ETCD_CERT_FILE:/certs/etcd-client.pem}"
+    cert_key_file: "${ETCD_CERT_KEY_FILE:/certs/etcd-client-key.pem}"
+    ca_file: "${ETCD_CA_FILE:/certs/etcd-ca.pem}"
+
+etcd-election@reconciler:
+  root_path: erda/component-leader/pipeline/reconciler
+
+etcd-election@gc:
+  root_path: erda/component-leader/pipeline/gc
```
Comment (sfwn) [inline conf/pipeline/pipeline.yaml:28]: which type of gc

**Review by sfwn — CHANGES_REQUESTED** [review]

**Comment by sfwn** [comment]
adjust your commits.

File: conf/pipeline/pipeline.yaml (hunk 2 above)
Comment (chengjoey) [reply conf/pipeline/pipeline.yaml:28]: i move all gc (namespace gc,database gc) to one election key, and pipeline reconciler,backup to one election key

File: modules/pipeline/initialize.go [hunk 3]
```diff
@@ -191,12 +189,14 @@ func (p *provider) do() (*httpserver.Server, error) {
 	// aop
 	aop.Initialize(bdl, dbClient, reportSvc)
 
-	// engine start after all dependencies done
-	engine.Start()
-	// handle cron related after engine started
-	if err := doCrondAbout(crondSvc, pipelineSvc); err != nil {
-		return nil, err
-	}
+	p.ReconcilerElection.OnLeader(func(ctx context.Context) {
```
Comment (sfwn) [inline modules/pipeline/initialize.go:192]: How about degrade from leader but pipeline still running? You have to release leader functions.

**Review by sfwn — CHANGES_REQUESTED** [review]

File: modules/pipeline/initialize.go (hunk 3 above)
Comment (sfwn) [reply modules/pipeline/initialize.go:192]: Use context.cancel or other methods.

Comment (chengjoey) [reply modules/pipeline/initialize.go:192]: use ctx.Done channel leave leader func

**Comment by chengjoey** [comment]
image
image

**Comment by chengjoey** [comment]
image
if ctx.done stop crond

File: modules/pipeline/pipengine/engine.go [hunk 4]
```diff
@@ -96,42 +99,6 @@ func (engine *Engine) WaitDBGC(pipelineID uint64, ttl uint64, needArchive bool)
 	engine.reconciler.WaitDBGC(pipelineID, ttl, needArchive)
 }
 
-const logPrefixContinueLoading = "continue load running pipelines"
-
-func (engine *Engine) continueLoadRunningPipelines() {
```
Comment (sfwn) [inline modules/pipeline/pipengine/engine.go:101]: Moved to where?

File: modules/pipeline/pipengine/reconciler/dbgc.go [hunk 5]
```diff
@@ -40,17 +40,25 @@ const (
...
 	r.doAnalyzedPipelineDatabaseGC(true)
 	r.doAnalyzedPipelineDatabaseGC(false)
-
 	r.doNotAnalyzedPipelineDatabaseGC(true)
 	r.doNotAnalyzedPipelineDatabaseGC(false)
 
-	time.AfterFunc(24*time.Hour, func() {
-		r.PipelineDatabaseGC()
-	})
+	ticker := time.NewTicker(24 * time.Hour)
```
Comment (sfwn) [inline modules/pipeline/pipengine/reconciler/dbgc.go:49]: close ticker defer

File: modules/pipeline/services/pipelinesvc/cron.go [hunk 6]
```diff
@@ -30,6 +31,11 @@ import (
 	"github.com/erda-project/erda/modules/pipeline/services/appsvc"
 	"github.com/erda-project/erda/modules/pipeline/spec"
 	"github.com/erda-project/erda/pkg/discover"
+	"github.com/erda-project/erda/pkg/loop"
+)
+
+const (
+	ReconcilerLeaderElectionPrefix = "/pipeline/leaderelection/do-cron"
```
Comment (sfwn) [inline modules/pipeline/services/pipelinesvc/cron.go:38]: where to use

File: modules/pipeline/services/pipelinesvc/cron.go [hunk 7]
```diff
@@ -187,3 +193,40 @@ func (s *PipelineSvc) UpgradePipelineCron(pc *spec.PipelineCron) error {
...
+	}
+	if err != nil {
+		logrus.Errorf("failed to reload crond from db (%v)", err)
+		return
+	}
+
+	// watch crond
+	go s.crondSvc.ListenCrond(ctx, s.RunCronPipelineFunc)
+
+	// 定时打印定时任务快照
```
Comment (sfwn) [inline modules/pipeline/services/pipelinesvc/cron.go:212]: use english

**Review by sfwn — CHANGES_REQUESTED** [review]

File: modules/pipeline/pipengine/engine.go (hunk 4 above)
Comment (chengjoey) [reply modules/pipeline/pipengine/engine.go:101]: image
No usages found in All Places, Its function is to import the running pipeline,engine.loadRunningPipelines() already carried out

File: modules/pipeline/services/pipelinesvc/cron.go (hunk 6 above)
Comment (chengjoey) [reply modules/pipeline/services/pipelinesvc/cron.go:38]: no usages,i deleted it

File: modules/pipeline/pipengine/engine.go (hunk 4 above)
Comment (chengjoey) [reply modules/pipeline/pipengine/engine.go:101]: i restored it, may use it later

**Review by sfwn — APPROVED** [review]

**Pull request merged**